Recheck for cloud taint before initializing nodes in cloud controller

tghartland opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

If a node creates an Update event before its Add event has been processed by the node controller, after the Add event has been processed the Update is processed and the node object received by the Update event is outdated and does not reflect that the node has already been initialized. Therefore the Update event initializes the node again.

With enough nodes that the Update event rate is higher than the rate at which the node controller can call initializeNode (taking into account the time spent making queries to the cloud), events start to pile up and a node which os added later than others will have its initial Add event at the end of the event queue, meaning it can take a long time for it to be processed. And while that node is waiting to be initialized, it is filling up the event queue with more Update events, which are also going to call initializeNode again.

By maintaining an internal map of nodes that have been initialized, outdated Update events for nodes that are known to have already been initialized can be discarded.

There is more detailed information in issue #75285

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #75285 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Fix bug where cloud-controller-manager initializes nodes multiple times
```

tghartland commented:
/sig cloud-provider
/sig scalability
/area cloudprovider
/area controller-manager

tghartland commented:
@andrewsykim

andrewsykim commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -204,8 +211,16 @@ func (cnc *CloudNodeController) UpdateCloudNode(_, newObj interface{}) {
+	cnc.initializedNodesMutex.Lock()
+	machineID, found := cnc.initializedNodes[node.Name]
+	cnc.initializedNodesMutex.Unlock()
+	alreadyInitialized := found && (machineID == node.Status.NodeInfo.MachineID)
```
I believe `NodeInfo.MachineID` is only set if cAdvisor is enabled, not sure if we want to be relying on cAdvisor behavior in this way here 🤔

tghartland replied:
Ah ok. Is there another way to uniquely identify a node like this?
Otherwise I think I can manage without it by removing a name from the map when it is eventually seen in an Update event where the node object appears initialized, and can be discarded for that reason from then on.

yastij commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -204,8 +211,16 @@ func (cnc *CloudNodeController) UpdateCloudNode(_, newObj interface{}) {
+	cnc.initializedNodesMutex.Lock()
+	machineID, found := cnc.initializedNodes[node.Name]
+	cnc.initializedNodesMutex.Unlock()
+	alreadyInitialized := found && (machineID == node.Status.NodeInfo.MachineID)
```
we might want to rely on providerID of the instance here cc @andrewsykim

andrewsykim replied:
The provider ID would not be set yet I believe

andrewsykim replied:
With the current state of the PR, I think it's possible for the provider ID to be not set 🤔

yastij replied:
on the kubelet we have: 
```golang
		if node.Spec.ProviderID == "" {
			node.Spec.ProviderID, err = cloudprovider.GetInstanceProviderID(context.TODO(), kl.cloud, kl.nodeName)
			if err != nil {
				return nil, err
			}
		}
```

Basically it relies on InstanceID call from instances interface, so as long as the cloud-provider implements this it should be fine. thoughts ?

andrewsykim replied:
IIRC, that code path is only for in-tree providers. This controller only runs out-of-tree.

yastij replied:
True, there's also some other ways:

- store on the map the `providerID` after initialization and use `cloudprovider.GetInstanceProviderID(context.TODO(), cnc.cloud, types.NodeName(curNode.Name))` for comparison (this might be expensive as need to make call to retrieve `instanceID()`)

- change the data structure to a set of `nodeName` and make sure we add the nodeName to the set at the of the initialization, when receiving an update we can check for presence on the set

tghartland replied:
I have just added some commits to change it to only track the node name in a set, and then remove the name from the set when it is safe to use the old behaviour of just checking the cloud taint on the event node object. This works in my tests.
For now I left in some logging that I want to remove before merging. The example output when grepping the logs for a specific node looks like
```
I0321 14:31:36.076005       1 node_controller.go:355] Successfully initialized node scaler-49-n7fhsowneckq-minion-49 with cloud provider, time=1.651611
I0321 14:31:36.076041       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:31:36.076049       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:31:36.076057       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:31:36.076062       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:01.424251       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:32:01.424255       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:01.424260       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:32:01.424265       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:31.030561       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:32:31.030565       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:32.694513       1 node_controller.go:222] Node scaler-49-n7fhsowneckq-minion-49 is in the recently initialized nodes set, will not re-process
I0321 14:32:32.694520       1 node_controller.go:230] Node scaler-49-n7fhsowneckq-minion-49 is being removed from the recently initialized set as its update events no longer have the cloud taint
I0321 14:32:32.694525       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:32.694551       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:32:51.330381       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
I0321 14:33:36.002248       1 node_controller.go:236] Nothing to do for node scaler-49-n7fhsowneckq-minion-49
```
Once the name is removed from the set it is no problem if the node is removed on the cloud side and a new node with the same name is added.

Thoughts on this solution?

tghartland replied:
Though without another property (machine / provider ID) to tell the difference between two nodes with the same name, there is the possibility that a node which is removed on the cloud side immediately after it is initialized in kubernetes (and therefore never sending the update event which would remove the name from the set) would not be removed from the set.
The Add event of the next node to be added with that name would still be processed, and the state would be back to normal after that, but if the Add fails and it is supposed to rely on the Update as a backup then that would not work.

yastij replied:
If the add fail we would still remove the nodeName from set no ?

yastij replied:
If the add fails there's no reason that we keep its nodeName on the set, so when receiving an update event we'd still initialize the node

tghartland replied:
The removal of names in the set is done when the Update events see the transition of cloud taint -> no cloud taint on the event node object. If the Add event initialization for the second node with the same name failed and it wasn't able to do the initialization on an Update event because the name already exists in the set, it would never be initialized so that transition would never be seen.

Maybe going back to a map and having the value in the map be the time that the name is added would be a way to check if a name should really still be tracked or not.

yastij replied:
@tghartland - if we are not able to initialize the node we shouldn't add it to the set, so if there's an error on the add we won't add the nodeName.
If we remove the node right after initialization and add it back, the add event will be processed and start initialization, with 2 possible outcomes I think:

- initialization is achieved, we add the nodeName from the set
- initialization failed, we remove the nodeName from the set if it exists

when receiving an update after the add event the set will behave as it should since if the initialization is done the nodeName would be on the set.

tghartland replied:
> initialization failed, we remove the nodeName from the set if it exists

In fact it doesn't even have to fail, if the name is in the set at the start of an Add event then it must have been left over from a previous node with the same name, and should be removed.
Commit fixing this: https://github.com/kubernetes/kubernetes/pull/75405/commits/7f12d7dcf75e74e19f601b9f9518165937b6e3c9
I think this takes care of any issues to do with a name not being removed properly.

yastij replied:
This would work too. more generally, seeing an add should clean the set from the nodeName

yastij commented:
/ok-to-test

yastij commented:
/assign
/priority important-soon

yastij commented:
/lgtm code-wise

yastij commented:
/lgtm

andrewsykim commented on pkg/controller/cloud/node_controller.go:
```diff
@@ -55,6 +57,12 @@ type CloudNodeController struct {
+	// Set of node names that have been initialized recently.
+	// Used to discard UpdateCloudNode events in which the received node object still
+	// has the cloud taint, even though the node has actually been initialized.
+	recentlyInitializedNodes      sets.String
```
Still wrapping my head around this PR so sorry in advance. I'm not entirely happy with the semantics around "recently initialized" 🤔 IMHO, it's a bit difficult to grok how this is used to ensure we're not initializing a node that is already being initialized, especially with the way it's used differently between OnAdd and OnUpdate events. Thoughts on this @cheftako?

tghartland replied:
Maybe I still haven't managed to explain properly. Does a picture with a timeline help? This is what (I think) is currently happening:

![node-init](https://user-images.githubusercontent.com/11710676/55387572-f0030900-5531-11e9-9b84-e3b26cb5fd82.png)
With the fix in this PR, all the update events are ignored because the node name is remembered from t=35 (after the initialization is first completed) so that the update events from before t=35 can be ignored. In this case the Update events can all be ignored and be cleared out from the event queue very quickly, allowing other nodes to be initialized. For example using the fix, in [these logs](https://gist.github.com/tghartland/1bdaffc718984d8f0faa1a02bdb53c13) there are many events for minion-41 that are only ignored because the name is remembered, which would otherwise have been re-processed.

When the update events are processed the actual Add event initialization was done in the recent past, not both happening at the same time, which is why I chose the name. If there is a shorter name that works I would change to that, since recentlyInitializedNodesMutex is a bit long.

I hope this makes things clearer.

yastij commented:
/cc @wojtek-t

andrewsykim commented:
/hold
/priority important-longterm

I think we need to be careful about the changes proposed here. The logic makes sense but I worry it complicates the controller logic significantly. I need a bit more time to think this through, I think this works but if there's a simpler way to accomplish this I would prefer it.

tghartland commented:
Since we ruled out querying the current state of the node on every update due to the number of requests it would create to the API, the only other thing I can think of is to check the status.conditions heartbeat time on the received node object and use that to discard events any that are older than a set amount of time. 25s or so would probably do it. The heartbeat is every 10s as well but I'm not sure how that would align with the the update events, so double and a bit extra puts us at 25s.

The Add events should always be processed, it's just the Update ones that should be checked. Even if a few nodes still do get initialized multiple times, as long as the number is greatly reduced then the exponential pile-up of events should be avoided.

I'll try this in practice and see how it works.

tghartland commented:
Based on the time difference between the last heartbeat of the received node object and the time that the event is processed, I added some logging for the age of the received node objects.
```go
--- a/pkg/controller/cloud/node_controller.go
+++ b/pkg/controller/cloud/node_controller.go
@@ -204,6 +204,16 @@ func (cnc *CloudNodeController) UpdateCloudNode(_, newObj interface{}) {
                return
        }
 
+       lastHeartbeat := metav1.Time{}
+       conditions := node.Status.Conditions
+       for _, cond := range conditions {
+               if cond.LastHeartbeatTime.After(lastHeartbeat.Time) {
+                       lastHeartbeat = cond.LastHeartbeatTime
+               }
+       }
+
+       klog.Infof("Update %s, age %.01f", node.Name, time.Since(lastHeartbeat.Time).Seconds())
+
        cloudTaint := getCloudTaint(node.Spec.Taints)
        if cloudTaint == nil {
                // The node has already been initialized so nothing to do.
```
Then with the [output logs](https://gist.github.com/tghartland/409bb145114e893f5a685ca842fb02bb) for a test scaling from 1 node to 50, I plot how the age of the received node object by the time it is processed increases over time.
![output54](https://user-images.githubusercontent.com/11710676/56221806-38aade00-606b-11e9-9584-a053f5365f4b.png)

There were a few nodes still being initialized at around 50 minutes, but once there was only one left the age started to drop and once the cloud-controller-manager stopped receiving Update events with the cloud taint for that one remaining node the entire queue was processed almost instantly (~10,000 events cleared in ~1 second).

yastij commented:
I think one of the drawbacks of this is that it is very dependent of your users configuration (node update frequency) so this would make it difficult to find an heuristic.

Also long-term, how this would play with the Lease API ? cc @wojtek-t

tghartland commented:
> very dependent of your users configuration (node update frequency)

That is a bit of a problem, yes. I've tried discarding events older than 25s and it does work, but I haven't found any way to find what the node update time is in order to set the age limit. It could be determined by looking at the heartbeat times of two sequential events for the same node, but that's definitely too hacky.

The only other idea I have is to replace the node informer event handlers with methods that immediately read the events into two slices, one for Add events and one for Update events. Then we would prioritise processing the Add events over the Update events in a processing loop that we run ourselves. This would minimise the number of Update events that are generated with the cloud taint still on them, preventing the [exponential growth](https://user-images.githubusercontent.com/11710676/54189291-c6663d00-44b1-11e9-9043-f313cd7ffed4.png) that is caused by Update events blocking Add events of the latest nodes to be registered.

tghartland commented:
https://github.com/kubernetes/client-go/blob/c34bf9c3f770582c59dc5871b9ed70b3f809ed82/tools/cache/fifo.go#L81-L93
Better than my last suggestion, reading events from the node informer into a FIFO queue should work. The ID used by the queue would be the node name, so that even if there have been 1 Add and 100 Updates for the node before any of them have been processed, or the node has been replaced with a completely different node with the same name, when it is popped from the FIFO it is the most recent version that is retrieved, and the node will not exist in the FIFO again until the next actual Update of the node.

This takes care of discarding unneeded events for us. There would not be any way to distinguish between processing Add and Update events any more, but the difference between AddCloudNode and UpdateCloudNode is minimal anyway.

Thoughts on this, @yastij, @andrewsykim?

tghartland commented:
The tests are going to complain a lot but I have force pushed with the FIFO queue. In practice it works, I'll fix all the tests later.
What do you think of this implementation?

andrewsykim commented:
@tghartland sorry for the delay. Wouldn't we still be pushing nodes with the taint into the queue while a node is being registered?

tghartland commented:
@andrewsykim That would happen yes, but the node would go to the end of the queue. As long as there are other nodes in the queue that have to be initialized before it, then another update will come from the node without the cloud taint and replace the tainted node object in the queue.

The worst case scenario is that some nodes might get initialized twice, but that's a lot better than what can happen currently.

tghartland commented:
/retest

yastij commented:
I agree with @andrewsykim - it seems like we'll queue events containing stale node objects @tghartland

tghartland commented:
@yastij the node objects in the queue can only be as stale as their latest Update event, up to 10 seconds at max (or whatever the update frequency is set to). To have anything less stale than this would require re-querying the node API for the current state upon processing the event, which is something we [already ruled out](https://github.com/kubernetes/kubernetes/issues/75285#issuecomment-472202047).

Without doing that, it is possible that some nodes get initialized two times (at maximum, and it would be rare), but I think that is as good as it gets.

andrewsykim commented:
@tghartland sorry for the delay in reviewing this, I think we should fix this, but just thinking through how to do this correctly.

andrewsykim commented:
cc @feiskyer

andrewsykim commented:
@tghartland I think queue is the right approach here but I think a larger refactor is going to be needed where we queue the events not the node object. See [sample-controller](https://github.com/kubernetes/sample-controller/blob/master/controller.go#L63-L237) for an example of what I mean. If we can guarantee that only 1 node event is processed at a time (process queue with threadiness=1), then we can fetch the latest node object before initialization (either from cache or directly from the API) and check the taint state before continuing.

strigazi commented:
cc @hogepodge

tghartland commented:
@andrewsykim It was my belief that even in the current system the node events are being processed only one at a time. For example in the logs I link to in [this comment](https://github.com/kubernetes/kubernetes/issues/75285#issuecomment-473279365), which has over 1000 node initializations, the minimum time between two log lines is 0.77 seconds, which is at the same order of magnitude as the average initialization (1 second). If the initializations were being processed in multiple threads, I would expect this value to be lower.

I will take a look at the sample controller, but I think the important aspect of the implementation I have here currently is that the cache.FIFO queue is never longer than the number of nodes; each node can only exist once in the queue (checked by node name). This ensures that even if the initialization time becomes large (cloud API is having issues and latency increases, for example), the queue will not grow out of control.

andrewsykim commented:
If we know only 1 node is being processed at a time, can we just fetch the latest node object at the beginning of `initializeNode` and recheck the node taint before continuing?

tghartland commented:
> If we know only 1 node is being processed at a time, can we fetch the latest node object at the beginning of initializeNode (get from client) and recheck the node taint before continuing?

We can, the question is will this cause too many API requests. The only time the request needs to be made is when the cloud taint is present, so when the cluster is static it will not be called. Even when adding nodes, it will not need to be used often, but it could be possible to have a scenario where:

* 100s of nodes are added to a cluster, or a 100s of nodes cluster is created
* the first node initialization blocks indefinitely for some reason (waiting on cloud API maybe)
* Update events with the cloud taint start stacking up in the queue
* when the first initialization unblocks, all the remaining nodes are successfully initialized
* there is then a backlog of thousands of Update events with the cloud taint which are all processed and immediately discarded, but each one creates a request to the kube API, all at once

Though maybe that wouldn't be too much damage, I'm not really sure.

This is why I like the cache.FIFO. There is no reason to hold on to more than one event for a particular node in the queue. Maybe combining both solutions is the best option?

andrewsykim commented:
> We can, the question is will this cause too many API requests. The only time the request needs to be made is when the cloud taint is present, so when the cluster is static it will not be called. Even when adding nodes, it will not need to be used often, but it could be possible to have a scenario where:

It's less requests than you think because we only re-check the taint only if an update event resource doesn't have the taint. Fwiw, a get request to the API server is cheaper than multiple requests to cloud APIs so the trade-off seems worth it if the solution is simple. If we think a get request is too expensive, we can even check the node cache from the lister, there's still a chance we get a stale object but possible still an optimization from what we had before. 

> This is why I like the cache.FIFO. There is no reason to hold on to more than one event for a particular node in the queue. Maybe combining both solutions is the best option?

As you mentioned, with the FIFO cache there's still a chance we would reinitialize a node, so not sure what the benefit of adding that complexity is over just re-checking state in cache.

andrewsykim commented:
Looks like we already get the latest state of the node [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L254-L257). If we're already making the API request let's just recheck the taint state there

tghartland committed:
Abort node initialization if cloud taint was already removed

If node events are received at a faster rate than they
can be processed then initialization for some nodes will
be delayed. Once they are eventually processed their cloud taint
is removed, but there may already be several update events
for those nodes with the cloud taint still on them already
in the event queue. To avoid re-initializing those nodes,
the cloud taint is checked for again after requesting
the current state of the node. If the cloud taint is no
longer on the node then nil is returned from the
RetryOnConflict, as an error does not need to be logged.

The logging for a successful initialization is also
moved inside the RetryOnConflict so that the early nil
return does not cause the aborted initialization to be
logged as a success.

andrewsykim commented:
> Also, getting the current node state should probably be moved to the top of the function passed to RetryOnConflict, yes? Otherwise that GCE code is still going to be run.

That sounds good to me. 

> It is not possible to return from the RetryOnConflict with nil, as that then prints the "Successfully initialized" line. We could return a custom error type and check for that to return from initializeNode without logging anything.

Let's move the log line into the end of the retry block?

tghartland commented:
> Let's move the log line into the end of the retry block?

Good idea. I'll sort out moving those changes into this PR's branch tomorrow, then remove the WIP status (and probably change the title as well).

andrewsykim commented:
> Good idea. I'll sort out moving those changes into this PR's branch tomorrow, then remove the WIP status (and probably change the title as well).

Sounds good, thanks for iterating @tghartland much appreciated!!

tghartland commented:
Moving getting the current state above the GCE specific part does introduce one complication which the tests caught.
```
=== RUN   TestGCECondition
--- FAIL: TestGCECondition (0.00s)
    node_controller_test.go:374: 
                Error Trace:    node_controller_test.go:374
                Error:          Not equal: 
                                expected: 2
                                actual  : 1
                Test:           TestGCECondition
                Messages:       No new conditions were added for GCE
    node_controller_test.go:383: 
                Error Trace:    node_controller_test.go:383
                Error:          Should be true
                Test:           TestGCECondition
                Messages:       Network Route Condition for GCE not added by external cloud initializer
```
What happens is that the `nodeutil.SetNodeCondition` modifies the node state, so when the request for the current node state is done first it is no longer the current state after setting the node condition. Then when the node is updated later, it removes the conditions that were added in the GCE part.

Either we should leave the GCE code at the top of RetryOnConflict, or the current state has to be refreshed after `SetNodeCondition`, like so:
```go
    err := clientretry.RetryOnConflict(UpdateNodeSpecBackoff, func() error {
            curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
            if err != nil {
                    return err 
            }   
 
            cloudTaint := getCloudTaint(curNode.Spec.Taints)
            if cloudTaint == nil {
                    // Node object received from event had the cloud taint but was outdated,
                    // the node has actually already been initialized.
                    return nil 
            }   
 
            // TODO(wlan0): Move this logic to the route controller using the node taint instead of condition
            // Since there are node taints, do we still need this?
            // This condition marks the node as unusable until routes are initialized in the cloud provider
            if cnc.cloud.ProviderName() == "gce" {
                    if err := nodeutil.SetNodeCondition(cnc.kubeClient, types.NodeName(node.Name), v1.NodeCondition{
                            Type:               v1.NodeNetworkUnavailable,
                            Status:             v1.ConditionTrue,
                            Reason:             "NoRouteCreated",
                            Message:            "Node created without a route",
                            LastTransitionTime: metav1.Now(),
                    }); err != nil {
                            return err 
                    }   
+
+                   var err error
+                   curNode, err = cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
+                   if err != nil {
+                           return err 
+                   }   
            }
```

yastij commented:
@tghartland @andrewsykim - Sorry for the delay

> But I'm not sure that it's a good idea to trust that the manually changed node will always be identical to what `SetNodeCondition` does. It does make the tests pass though.

I think this can be avoided using patch as we'll be sending only a diff change, this would avoid having conflicts on the node object. There's a broader issue with context here #56350

tghartland commented:
@yastij do you mean using patch instead of update here?
https://github.com/kubernetes/kubernetes/blob/b9bde60cfd01cf657fec0dadba826c57aec4dcf9/pkg/controller/cloud/node_controller.go#L306-L308
Because `SetNodeCondition` has this line
https://github.com/kubernetes/kubernetes/blob/b9bde60cfd01cf657fec0dadba826c57aec4dcf9/pkg/util/node/node.go#L154
which means that the condition you pass to the function isn't exactly what gets set on the node. The only way to get the exact condition that is set on the node is to request the node state from the API again. Without doing that the heartbeat time would be changed when sending the update.

andrewsykim commented:
@tghartland on second thought, let's recheck the node after the setting the GCE node condition. I don't see any harm in re-adding the NoRoute condition there, that should be idempotent anyways.

andrewsykim commented:
/approve
/lgtm

Thanks for iterating @tghartland!

andrewsykim commented:
Can you update the release note please, maybe something like "fix bug where cloud-controller-manager initializes nodes multiple times"

tghartland commented:
@andrewsykim I've changed the note to what you suggested.

andrewsykim commented:
/hold cancel
